File: 1_AudioRecoder/AudioRecorderDemo/libaudio/src/main/java/com/baishengye/libaudio/recorder/AudioChunk.java
package com.baishengye.libaudio.recorder;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

/**
 * 音频数据包装器
 */
public interface AudioChunk {
    // 获取最大峰值(振幅)
    double maxAmplitude();

    // 获取byte类型数据
    byte[] toBytes();

    // 获取short类型数据
    short[] toShorts();


    abstract class AbstractAudioChunk implements AudioChunk {
        private static final double REFERENCE = 0.6;

        @Override
        public double maxAmplitude() {
            int nMaxAmp = 0;
            for (short sh : toShorts()) {
                if (sh > nMaxAmp) {
                    nMaxAmp = sh;
                }
            }
            if (nMaxAmp > 0) {
                return Math.abs(20 * Math.log10(nMaxAmp / REFERENCE));
            } else {
                return 0;
            }
        }
    }

    /**
     * byte类型数据包装器
     */
    class Bytes extends AbstractAudioChunk {
        private final byte[] bytes;

        Bytes(byte[] bytes) {
            this.bytes = bytes;
        }

        @Override
        public byte[] toBytes() {
            return bytes;
        }

        @Override
        public short[] toShorts() {
            short[] shorts = new short[bytes.length / 2];
            ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN).asShortBuffer().get(shorts);
            return shorts;
        }
    }
}

File: 1_AudioRecoder/AudioRecorderDemo/libaudio/src/main/java/com/baishengye/libaudio/config/AudioChunk.kt
package com.baishengye.libaudio.config

import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.math.abs
import kotlin.math.log10

/**
 * 音频数据包装器
 */
interface AudioChunk {
    // 获取最大峰值(振幅)
    fun maxAmplitude(): Double

    // 获取byte类型数据
    fun toBytes(): ByteArray

    // 获取short类型数据
    fun toShorts(): ShortArray
    abstract class AbstractAudioChunk : AudioChunk {
        override fun maxAmplitude(): Double {
            var nMaxAmp = 0
            for (sh in toShorts()) {
                if (sh > nMaxAmp) {
                    nMaxAmp = sh.toInt()
                }
            }
            return if (nMaxAmp > 0) {
                abs(20 * log10(nMaxAmp / REFERENCE))
            } else {
                0.0
            }
        }

        companion object {
            private const val REFERENCE = 0.6
        }
    }

    /**
     * byte类型数据包装器
     */
    class Bytes(private val bytes: ByteArray) : AbstractAudioChunk() {
        override fun toBytes(): ByteArray {
            return bytes
        }

        override fun toShorts(): ShortArray {
            val shorts = ShortArray(bytes.size / 2)
            ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN).asShortBuffer()[shorts]
            return shorts
        }
    }
}